### DISCUSSION TITLE
Dynamic Loading questions

### DISCUSSION THREAD

**Discussion by TobyKLight** [cables-gl/cables_docs#617] [Answers & Questions] [answered]
I'm working on a project that will use a lot of high quality textures. 
Usecase is something like a gallery of thumbnails and when you click on one you get the high quality version. 
There will be far more textures than I want to put in VRAM or download to start the patch, so I need to do some dynamic loading.
I will use Texture and TextureArrayLoaderFromArray ops. 
I will run this on a private webserver. 
I saw from searching the discord that cables only loads assets as they are 'used', and there seems to be several ops that talk about preloading, which I take to mean preloading to VRAM. 
But overall I still have a fuzzy picture how texture loading and unloading works so I hoped you would indulge me with a few questions. 

* When does cables.gl download a texture from the server? 

* When does cables.gl load a texture into VRAM? (EG As soon as the texture op has an input file/string that is not yet loaded? OR when the texture is used by a triggered material? OR when that material is used on a mesh?) 

* If it's the case that textures have to be used by downstream ops to load, do the downstream ops have to have active trigger from mainloop to start loading textures?  

* When does cables.gl unload a texture from VRAM? 

* Does an unloaded texture stay available in some kind of client cache, so it can be loaded to VRAM again without download? 

* Does this work the same for TextureArrayLoaderFromArray? It has loader in the name but just to confirm: does it do the loading as soon as it has unloaded URLs on it's input? or does it also wait for textures to be used by a downstream op? Do all textures in the array get loaded when one is used? 

* What does the 'caching' toggle do on TextureArrayLoaderFromArray? 

* What does the 'Asset in patch' toggle do on TextureArrayLoaderFromArray?

**Comment by pandrr** [comment, marked as answer]
`When does cables.gl download a texture from the server?`

loading starts automatically on patch load or when you change the filename in the texture op (via clicking or changing a string that is connected)

`When does cables.gl load a texture into VRAM? `
as soon the filename is there or changed (AND the active toggle is on)

`If it's the case that textures have to be used by downstream ops to load, do the downstream ops have to have active trigger from mainloop to start loading textures?`
no, they are independent. the texture op oputputs a transparent texture until the real file is loaded. 

`When does cables.gl unload a texture from VRAM?`
- when deleting the op
- active port is set to false
- when changing the filename to a different texture (e.g. smaller size)

`Does an unloaded texture stay available in some kind of client cache, so it can be loaded to VRAM again without download?
`
depends on the browser/server caching settings. check network dev tools to analyze what is loaded from the server etc. cables does not cache those things, if you have multiple texture ops with the same file it will use multiple times of ram (use variables)

` Does this work the same for TextureArrayLoaderFromArray? It has loader in the name but just to confirm: does it do the loading as soon as it has unloaded URLs on it's input? or does it also wait for textures to be used by a downstream op? Do all textures in the array get loaded when one is used?`
all items in the array are agaion references to small transparent default textures


`    What does the 'caching' toggle do on TextureArrayLoaderFromArray?`
when turned off it adds a random number to the query parameter so browsers dont cache them

`    What does the 'Asset in patch' toggle do on TextureArrayLoaderFromArray?`
it helps with loading patches from different servers and when exporting the urls need to be changed etc.

i have an item on my todo list to rework the texturearray loading ops at one point and update them to the current workflows (e.g. they dont add to the loading status etc..) let me know if there are problems!

**Comment by TobyKLight** [comment]
Thanks so much for the answers ❤️.
 very useful.
